aten/src/ATen/native/cpu/GridSamplerKernel.cpp

```diff
@@ -1201,63 +1222,94 @@ grid_sampler_2d_backward_cpu_kernel_impl(const Tensor& grad_output_,
   // contiguous can greatly simplify this code.
   auto grad_output = grad_output_.contiguous();
 
-  auto grad_input = at::zeros_like(input, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  // If `input` gradient is not required, we skip computing it -- not needing to create
+  // the tensor to hold the gradient can markedly increase performance. (`grid` gradient
+  // is always computed.)
+  auto input_requires_grad = output_mask[0];
+
+  Tensor grad_input;
+  if (input_requires_grad) {
+    grad_input = at::zeros_like(input, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  }
   auto grad_grid = at::empty_like(grid, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
   auto N = input.size(0);
   auto spatial_size = grid.size(1) * grid.size(2);
   auto grain_size = spatial_size == 0 ? (N + 1)
                                       : at::divup(at::internal::GRAIN_SIZE, spatial_size * 10 /* 2d * 5 tensors*/);
 
-#define HANDLE_CASE(interp, padding, align_corners)                              \
+#define GINP_SLICE_PTR_true auto gInp_slice = gInp_acc[n]; auto gInp_slice_ptr = &gInp_slice;
+#define GINP_SLICE_PTR_false TensorAccessor<scalar_t, 3>* gInp_slice_ptr = nullptr;
+#define GINP_SLICE_PTR(input_requires_grad) GINP_SLICE_PTR_##input_requires_grad
```

Comment:
Why do you need these macros? an if statement would work as well no?

Reply:
As now implemented, `gInp_acc` is here only in scope for `input_requires_grad` is true.

It could be changed so that it's always declared and then use an if statement instead of the macros. But in that case, I think it will need to be changed to a pointer, as I don't think one can do `auto gInp_acc = grad_input.accessor<scalar_t, 4>();` when `grad_input` is undefined tensor (but the pointer could be made null in that case). Would the pointer approach be preferable to the macros (or perhaps there's other, better ways)?